src/index/src/bloom_filter/creator/finalize_segment.rs

```diff
@@ -0,0 +1,315 @@
+            .fetch_add(memory_diff, Ordering::Relaxed);
+
+        // Add the finalized Bloom filter to the in-memory storage.
+        self.in_memory.push(fbf);
+
+        // Flush to disk if necessary.
+
+        // Do not flush if memory usage is too low.
+        if self.memory_usage < MIN_MEMORY_USAGE_THRESHOLD {
+            return Ok(());
+        }
+
+        // Check if the global memory usage exceeds the threshold and flush to disk if necessary.
+        if let Some(threshold) = self.global_memory_usage_threshold {
+            let global = self.global_memory_usage.load(Ordering::Relaxed);
+
+            if global > threshold {
+                self.flush_in_memory_to_disk().await?;
+
+                self.global_memory_usage
+                    .fetch_sub(self.memory_usage, Ordering::Relaxed);
+                self.memory_usage = 0;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Drains the storage and returns a stream of finalized Bloom filter segments.
+    pub async fn drain(
+        &mut self,
+    ) -> Result<Pin<Box<dyn Stream<Item = Result<FinalizedBloomFilterSegment>> + '_>>> {
+        // FAST PATH: only memory
+        if self.intermediate_file_id_counter == 0 {
+            return Ok(Box::pin(stream::iter(self.in_memory.drain(..).map(Ok))));
+        }
+
+        // SLOW PATH: memory + disk
+        let mut on_disk = self
+            .intermediate_provider
+            .read_all(&self.intermediate_prefix)
+            .await
+            .context(IntermediateSnafu)?;
+        on_disk.sort_unstable_by(|x, y| x.0.cmp(&y.0));
+
+        let streams = on_disk.into_iter().map(|(_, mut reader)| {
+            async_stream::try_stream! {
+                let mut size_buf = [0u8; 16];
+                let mut bloom_filter_buf = vec![];
+
+                while reader.read_exact(&mut size_buf).await.is_ok() {
+                    let element_count = u64::from_le_bytes(size_buf[..8].try_into().unwrap()) as usize;
+                    let size = u64::from_le_bytes(size_buf[8..].try_into().unwrap()) as usize;
+
+                    bloom_filter_buf.clear();
+                    bloom_filter_buf.resize(size, 0);
+                    reader.read_exact(&mut bloom_filter_buf).await.context(IoSnafu)?;
+
+                    yield FinalizedBloomFilterSegment {
+                        bloom_filter_bytes: bloom_filter_buf.clone(),
+                        element_count,
+                    };
+                }
+            }
+        });
+
+        let in_memory_stream = stream::iter(self.in_memory.drain(..)).map(Ok);
+        Ok(Box::pin(
+            stream::iter(streams).flatten().chain(in_memory_stream),
+        ))
+    }
+
+    /// Flushes the in-memory Bloom filters to disk.
+    ///
+    /// # Format
+    ///
+    /// [ elem count ][    size    ][ bloom filter ][ elem count ][    size    ][ bloom filter ]...
+    /// |<- u64 LE ->||<- u64 LE ->||<- bf bytes ->||<- u64 LE ->||<- u64 LE ->||<- bf bytes ->|...
```

Comment:
would a `version` number to added to file header too? Just in case future update might change file format?